pilot/pkg/proxy/envoy/v1/gateway.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+
+	"go.uber.org/zap"
+)
+
+func buildGatewayHTTPListeners(mesh *meshconfig.MeshConfig,
+	configStore model.IstioConfigStore, node model.Node) (Listeners, error) {
+
+	gateways, err := configStore.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		return nil, fmt.Errorf("listing gateways: %s", err)
+	}
+
+	if len(gateways) == 0 {
+		log.Debug("no gateways for router", zap.String("node", node.ID))
+		return Listeners{}, nil
+	}
+
+	gateway := &routing.Gateway{}
+	for _, spec := range gateways {
+		err := model.MergeGateways(gateway, spec.Spec.(*routing.Gateway))
+		if err != nil {
+			return nil, fmt.Errorf("merge gateways: %s", err)
+		}
+	}
+
+	listeners := make(Listeners, 0, len(gateway.Servers))
+	for _, server := range gateway.Servers {
+		// TODO: TCP
+
+		// build physical listener
+		physicalListener := buildPhysicalGatewayListener(mesh, node, configStore, server)
+		if physicalListener == nil {
+			continue // TODO: add support for all protocols
+		}
+
+		listeners = append(listeners, physicalListener)
+	}
+
+	return listeners.normalize(), nil
+}
+
+func buildPhysicalGatewayListener(
+	mesh *meshconfig.MeshConfig,
+	node model.Node,
+	config model.IstioConfigStore,
+	server *routing.Server,
+) *Listener {
+
+	opts := buildHTTPListenerOpts{
+		mesh:             mesh,
+		node:             node,
+		nodeInstances:    nil, // only required to support deprecated mixerclient behavior
+		routeConfig:      nil,
+		ip:               WildcardAddress,
+		port:             int(server.Port.Number),
+		rds:              strconv.Itoa(int(server.Port.Number)),
+		useRemoteAddress: true,
+		direction:        IngressTraceOperation,
+		outboundListener: false,
+		store:            config,
+	}
+
+	switch strings.ToUpper(server.Port.Protocol) {
+	case "HTTP":
+		return buildHTTPListener(opts)
+	case "HTTPS":
+		listener := buildHTTPListener(opts)
+		listener.SSLContext = tlsToSSLContext(server.Tls, server.Port.Protocol)
+		return listener
+	case "GRPC", "HTTP2":
+		listener := buildHTTPListener(opts)
```

Comment:
Nit. This can include all 3 forms http, http2, gRPC. I am unsure how people will interpret https in a gateway. If they think tls termination (I think in tls we have a tls mode that says passthrough, mutual, simple). If it’s passthrough, we simply treat the listener as a tcp proxy with SNI. If it’s non pass through (that check is missing btw), then we just have to check if the protocol is tcp or a http variant. If latter, setup a http listener and ssl context.